### PR TITLE
Make storage network calls non-retriable for throttling errors

### PR DESCRIPTION
**Pull request by pragya91**
**Issue:**
When server is throttled, it returns 429 error code along with retryAfter time value. In static workload, if retry-after is respected everything should work smoothly. However the workload is rarely static, i.e. there are always new clients, or clients who are issuing multiple request evert x minutes, which worsens the throttling and could result in clients just staying in an unrecoverable state.

**Solution:**
If a ThrottlingError is thrown when calling any of the ODSP storage services, FF will not perform retries if the host policy object says `disableRetriesOnStorageThrottlingError: true`. The throttling error will eventually be bubbled up to the host app and where the app decide the retry cadence.

The code change is done in `epochTracker `s `fetchCore` method, since that is the one common point where all the storage network calls are made. 

[AB#7056](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7056)

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspError.ts [hunk 1]
```diff
@@ -28,10 +28,9 @@ export function errorObjectFromSocketError(
 			undefined, // response from http request
 			socketError.error
 				? JSON.stringify({ error: socketError.error }, getCircularReplacer())
-				: undefined, // responseText
+				: undefined, // responseText,
+			{ odspError: true, relayServiceError: true },
```
Comment (NicholasCouri) [inline packages/drivers/odsp-driver/src/odspError.ts:32]: I thought we wanted the storage one instead, not the push one

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 2]
```diff
@@ -324,6 +324,18 @@ export function createOdspNetworkError(
 				);
 				break;
 			}
+		case 429: // Throttling error
+			// TODO: check !FeatureFlag value here to opt-out of this behavior.
+			if(props.storageServiceError){
```
Comment (NicholasCouri) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:329]: >storageServiceError [](http://example.com/codeflow?start=12&length=19)

if it is not a storage error, ie, push errors, we do want the default behavior below, isn't it ?

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 3]
```diff
@@ -324,6 +324,18 @@ export function createOdspNetworkError(
...
+				// Do not perform retries if storage service has throttled, unless opted-in by the client app.
```
Comment (NicholasCouri) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:330]: > Do not perform retries if storage service has throttled, unless opted-in by the client app. [](http://example.com/codeflow?start=6&length=92)

The default behavior should be the retriable attempts - shouldn't it ?

File: packages/utils/odsp-doclib-utils/api-report/odsp-doclib-utils.api.md [hunk 4]
```diff
@@ -172,7 +173,7 @@ export function putAsync(url: string, authRequestInfo: IOdspAuthRequestInfo): Pr
 export function refreshTokens(server: string, scope: string, clientConfig: IClientConfig, tokens: IOdspTokens): Promise<IOdspTokens>;
 
 // @internal
-export function throwOdspNetworkError(errorMessage: string, statusCode: number, response: Response, responseText?: string, props?: ITelemetryProperties): never;
+export function throwOdspNetworkError(errorMessage: string, statusCode: number, response: Response, responseText?: string, props?: ITelemetryProperties, hostPolicy?: HostStoragePolicy): never;
```
Comment (pragya91) [inline packages/utils/odsp-doclib-utils/api-report/odsp-doclib-utils.api.md:176]: Ignore this change. The signature is incorrect.

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 2 above)
Comment (pragya91) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts]: In the code I am actually letting the default condition handle the relay service scenario. Is you recommendation to explicitly add logic for non-storage errors as well within the case 429?

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 5]
```diff
@@ -290,27 +292,39 @@ export class EpochTracker implements IPersistedFileCache {
...
+								error.message,
+								redirectUrl,
+								{ driverVersion, redirectLocation },
+							);
+							locationRedirectionError.addTelemetryProperties(
+								error.getTelemetryProperties(),
+							);
+							throw locationRedirectionError;
+						}
+					}else if(error.errorType == OdspErrorTypes.throttlingError && this.hostPolicy?.disableRetriesOnStorageThrottlingError){
```
Comment (NicholasCouri) [inline packages/drivers/odsp-driver/src/epochTracker.ts:313]: > == [](http://example.com/codeflow?start=29&length=3)

===

**Review by NicholasCouri — APPROVED** [review]

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 6]
```diff
@@ -298,27 +300,48 @@ export class EpochTracker implements IPersistedFileCache {
...
-					error.errorType === OdspErrorTypes.fileNotFoundOrAccessDeniedError
-				) {
-					const redirectLocation = (error as IOdspErrorAugmentations).redirectLocation;
-					if (redirectLocation !== undefined) {
-						const redirectUrl: IOdspResolvedUrl = patchOdspResolvedUrl(
-							this.fileEntry.resolvedUrl,
-							redirectLocation,
-						);
-						const locationRedirectionError = new LocationRedirectionError(
+				if (isFluidError(error)) {
```
Comment (agarwal-navin) [inline packages/drivers/odsp-driver/src/epochTracker.ts:303]: Update the comment above to include the new behavior.

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 7]
```diff
@@ -298,27 +300,48 @@ export class EpochTracker implements IPersistedFileCache {
...
+					} else if (
+						error.errorType === OdspErrorTypes.throttlingError &&
+						this.hostPolicy?.disableRetriesOnStorageThrottlingError
+					) {
+						const nonRetriableThrottlingError = new NonRetryableError(
 							error.message,
-							redirectUrl,
-							{ driverVersion, redirectLocation },
+							OdspErrorTypes.throttlingError,
+							// retryAfterMs is still provided so that client app can can still use it.
```
Comment (agarwal-navin) [inline packages/drivers/odsp-driver/src/epochTracker.ts:329]: How is the host supposed to use this to decide the retry cadence as mentioned in the description?

File: packages/drivers/odsp-driver/src/test/epochTests.spec.ts [hunk 8]
```diff
@@ -393,4 +396,69 @@ describe("Tests for Epoch Tracker", () => {
 		}
 		assert.strictEqual(success, false, "Fetching should not succeed!!");
 	});
+
+	it("Checks throttling errors non-retriable when disableRetriesOnStorageThrottlingError=true", async () => {
```
Comment (agarwal-navin) [inline packages/drivers/odsp-driver/src/test/epochTests.spec.ts:400]: nit:
```suggestion
	it("Checks throttling errors are non-retriable when disableRetriesOnStorageThrottlingError=true", async () => {
```

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 9]
```diff
@@ -298,27 +300,48 @@ export class EpochTracker implements IPersistedFileCache {
...
+							{
+								retryAfterMs:
```
Comment (agarwal-navin) [inline packages/drivers/odsp-driver/src/epochTracker.ts:331]: Why not use `retryAfterSeconds` itself as the property name instead to introducing a new one?

**Comment by agarwal-navin** [comment]
> The throttling error will eventually be bubbled up to the host app and where the app decide the retry cadence.

It may be worth writing an end-to-end test to validate this happens and that hosts can actually reconnect as per their cadence.

File: packages/drivers/odsp-driver/src/epochTracker.ts (hunk 7 above)
Comment (pragya91) [reply packages/drivers/odsp-driver/src/epochTracker.ts:331]: My understanding is that this error gets bubbled up to the host app. The host can extract the error props (if they wish to use it) and make a reconnection request taking retryAfter into account if they choose to do so.

**Pull request merged**